src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -104,7 +104,18 @@ public ParseResult ParseMessage(ReadableBuffer buffer, out ReadCursor consumed,
                         consumed = lineEnd;
                         break;
                     case InternalParseState.ReadEndOfMessage:
-                        if (_data.Count > 0)
+                        if (_data.Count == 1)
+                        {
+                            if (_messageType == MessageType.Binary)
+                            {
+                                payload = MessageFormatUtils.DecodePayload(_data[0]);
```

Comment:
Don't do the decoding in the if branches. In the `if` branches, just assign `payload` to the single buffer that represents all the data we got from the client. Then **outside** this `if`, do another `if` on message type and decode in there. That way you only have to call `DecodePayload` once and there's a simple flow of operations:

1. Get the payload into a single buffer (via copying if necessary)
2. Decode the buffer **if necessary**
3. Put the buffer into a `Message` struct and return it.